doc/reference/shell/index.rst

```diff
@@ -542,6 +542,132 @@ are :c:macro:`SHELL_DEFINE` arguments.
 	RTT (:option:`CONFIG_LOG_BACKEND_RTT`), which are available earlier
 	during system initialization.
 
+Resources usage
+***************
+
+The following is an indication of the consumption of Flash and RAM by each
+shell feature. These features can be deactivated as needed to save CPU
+resources. The values were obtained by compiling on Cortex-M33.
+
++------------------+-----------+---------+--------------------------------------------------------------------+
+| Feature Name      | Flash [B] | RAM [B] |  Comment                                                          |
++==================+===========+=========+====================================================================+
+|  Shell           | 7012      | 1944    | | Shell module with very basic functionality.                      |
```

Comment:
Tables like this are going to get out of date really fast, and unless there's a way to automatically generate them I don't think they belong in the documentation.

I personally wouldn't keep this as documentation, but I would put each feature set into the shell test, to verify builds continue to work with the different combinations.  Just build, no need to have special tests for the different combinations---though there should be tests to confirm the behavior resulting from disabling a feature, where that's feasible.

The artifacts resulting could be used to calculate the impact of the changes.  If you want you could document the observed relative changes as comments in the testcase.yaml file so people can verify them over time without making people expect them to be accurate.

Reply:
Fair point.
I can rework tests and integrate them with the documentation but I think as a separate PR. This looks like a major piece of work for me.
What do you think?

Reply:
I personally think the various configurations should be build-tested to ensure they continue to work as intended, and that should be done here.  If nobody agrees I'll yield.

But I don't think it's worth integrating the analysis of size impact into the user documentation, certainly not in this PR.

Reply:
Ah; for the documentation part I was only looking for something like this:
```
diff --git a/doc/reference/shell/index.rst b/doc/reference/shell/index.rst
index 312b01316b..1d89cd3711 100644
--- a/doc/reference/shell/index.rst
+++ b/doc/reference/shell/index.rst
@@ -32,6 +32,14 @@ interaction is required. This module is a Unix-like shell with these features:
 * Support for meta keys.
 * Kconfig configuration to optimize memory usage.
 
+Some of these features have a significant impact on RAM and flash usage,
+but many can be disabled when not needed.  A configuration that should
+produce the minimum useful feature set is in
+:zephyr_file:`samples/subsys/shell/shell_module/prj_minimal.conf`:
+
+.. literalinclude:: ../../../samples/subsys/shell_module/prj_minimal.conf
+   :langage: text                  
+
 The module can be connected to any transport for command input and output.
 At this point, the following transport layers are implemented:
```

(Until I looked I wasn't aware of the ability to include the config file into the documentation.  The content above is based on in-tree material but I haven't tested it to see what it looks like.  Paths and rST syntax may be incorrect.  It may also be better placed somewhere after the overall description.)

The desire to have the test cases that you documented build-tested was based on https://github.com/zephyrproject-rtos/zephyr/pull/30116#issuecomment-730326468 and taking advantage of the distinct configurations you identified in your space analysis update.  I think retaining the existing minimal configuration, while adding configurations to individually disable the feature sets you identified, would be helpful both to users and for ensuring the flags continue to work (there are no missed dependencies).  You should be able to use `extra_args: OVERLAY_CONFIG="disable_foo.conf"` in `testcase.yaml` to separate the build tests.